library/src/main/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapter.java

```diff
@@ -26,183 +27,299 @@
+        extends RecyclerView.Adapter<VH> {
+
+    protected static final int TYPE_HEADER = -2;
+    protected static final int TYPE_ITEM = -1;
+    protected static final int TYPE_FOOTER = -3;
+
+    private H header;
+    private List<T> items = Collections.EMPTY_LIST;
+    private F footer;
+    private boolean showFooter = true;
+
+    //<editor-fold desc="seperate onCreateViewHolder">
+    /**
+     * Invokes onCreateHeaderViewHolder, onCreateItemViewHolder or onCreateFooterViewHolder methods
+     * based on the view type param.
+     */
+    @Override
+    public final VH onCreateViewHolder(ViewGroup parent, int viewType) {
+        VH viewHolder;
+        if (isHeaderType(viewType)) {
+            viewHolder = onCreateHeaderViewHolder(parent, viewType);
+        } else if (isFooterType(viewType)) {
+            viewHolder = onCreateFooterViewHolder(parent, viewType);
+        } else {
+            viewHolder = onCreateItemViewHolder(parent, viewType);
+        }
+        return viewHolder;
+    }
+
+    protected VH onCreateHeaderViewHolder(ViewGroup parent, int viewType) {
+        return null;
+    }
+
+    protected abstract VH onCreateItemViewHolder(ViewGroup parent, int viewType);
+
+    protected VH onCreateFooterViewHolder(ViewGroup parent, int viewType) {
+        return null;
+    }
+    //</editor-fold>
+
+    //<editor-fold desc="seperate onBindViewHolder">
+    /**
+     * Invokes onBindHeaderViewHolder, onBindItemViewHolder or onBindFooterViewHOlder methods based
+     * on the position param.
+     */
+    @Override
+    public final void onBindViewHolder(VH holder, int position) {
+        if (isHeaderPosition(position)) {
+            onBindHeaderViewHolder(holder, position);
+        } else if (isFooterPosition(position)) {
+            onBindFooterViewHolder(holder, position);
+        } else {
+            onBindItemViewHolder(holder, position);
+        }
+    }
+
+    protected void onBindHeaderViewHolder(VH holder, int position) {
+    }
+
+    protected abstract void onBindItemViewHolder(VH holder, int position);
+
+    protected void onBindFooterViewHolder(VH holder, int position) {
+    }
+    //</editor-fold>
+
+    //<editor-fold desc="seperate onFailedToRecycleView">
+    @Override
+    public final boolean onFailedToRecycleView(VH holder) {
+        int position = holder.getAdapterPosition();
+        boolean ret;
+
+        if (isHeaderPosition(position)) {
+            ret = onFailedToRecycleHeaderView(holder);
+        } else if (isFooterPosition(position)) {
+            ret = onFailedToRecycleFooterView(holder);
+        } else {
+            ret = onFailedToRecycleItemView(holder);
+        }
+
+        return ret;
+    }
+
+    protected boolean onFailedToRecycleHeaderView(VH holder) {
+        return false;
+    }
+
+    protected boolean onFailedToRecycleItemView(VH holder) {
+        return false;
+    }
+
+    protected boolean onFailedToRecycleFooterView(VH holder) {
+        return false;
+    }
```

Comment:
We don't used to use this part of the ``RecyclerView`` API. If I'm not wrong we can remove these four methods:

* ``onFailedToRecycleView``
* ``onFailedToRecycleHeaderView``
* ``onFailedToRecycleItemView``
* ``onFailedToRecycleFooterView``

Remember to update the README.md with these changes.